File: frontend/scripts/check-user-facing-text.mjs
#!/usr/bin/env node

import { createHash } from 'node:crypto'
import { readdir, readFile } from 'node:fs/promises'
import { extname, relative, resolve } from 'node:path'
import ts from 'typescript'

const frontendRoot = resolve(process.cwd())
const repositoryRoot = resolve(frontendRoot, '..')

function fingerprint(value) {
  return `sha256:${createHash('sha256').update(value).digest('hex')}`
}

function literalValue(node, sourceFile) {
  if (ts.isStringLiteral(node) || ts.isNoSubstitutionTemplateLiteral(node)) return node.text
  if (node.kind === ts.SyntaxKind.JsxText) {
    return node.getText(sourceFile).replace(/\s+/g, ' ').trim()
  }
  if (ts.isTemplateExpression(node)) return node.getText(sourceFile)
  return null
}

function isLiteral(node) {
  return ts.isStringLiteral(node)
    || ts.isNoSubstitutionTemplateLiteral(node)
    || node.kind === ts.SyntaxKind.JsxText
    || ts.isTemplateExpression(node)
}

function isResourcePath(path) {
  return path === 'frontend/src/lib/resources.ts'
    || path.startsWith('frontend/src/resources/')
}

function isTestPath(path) {
  return path.includes('/test/')
    || path.includes('/__tests__/')
    || /\.(?:test|spec)\.[cm]?[jt]sx?$/.test(path)
}

function isMachineContract(node, value) {
  if (value.startsWith('/')) return true
  const parent = node.parent
  if (ts.isPropertyAssignment(parent)) {
    const name = parent.name.getText().replace(/^['"]|['"]$/g, '').toLowerCase()
    return ['code', 'route', 'path', 'callbackdata', 'command'].includes(name)
  }
  return false
}

function isTelemetry(node) {
  let current = node.parent
  while (current && !ts.isStatement(current)) {
    if (ts.isCallExpression(current) && /^(?:console|logger)\./.test(current.expression.getText())) {
      return true
    }
    current = current.parent
  }
  return false
}

async function walk(directory) {
  const files = []
  for (const entry of (await readdir(directory, { withFileTypes: true }))
    .sort((left, right) => left.name.localeCompare(right.name))) {
    const path = resolve(directory, entry.name)
    if (entry.isDirectory()) {
      if (!['node_modules', 'dist', 'dist-catalog'].includes(entry.name)) files.push(...await walk(path))
    } else if (['.ts', '.tsx'].includes(extname(entry.name))) {
      files.push(path)
    }
  }
  return files
}

async function readEntries(path) {
  const document = JSON.parse(await readFile(path, 'utf8'))
  return document.entries ?? []
}

export async function scanUserFacingText({
  sourceRoot = resolve(frontendRoot, 'src'),
  exceptionsPath = resolve(
    repositoryRoot,
    'scripts/harness/config/user-facing-text-inventory-index/scanner-exceptions.json',
  ),
  allowlistPath = resolve(
    repositoryRoot,
    'scripts/harness/config/user-facing-text-allowlist.json',
  ),
  includeTests = false,
} = {}) {
  const exceptions = await readEntries(exceptionsPath)
  const allowlist = await readEntries(allowlistPath)
  const acceptedExceptions = new Map(exceptions
    .filter((entry) => entry.path.startsWith('frontend/'))
    .map((entry) => [`${entry.path}\0${entry.fingerprint}`, entry]))
  const acceptedAllowlist = new Map(allowlist
    .filter((entry) => entry.path.startsWith('frontend/'))
    .map((entry) => [`${entry.path}\0${entry.fingerprint}`, entry]))
  const seenExceptions = new Set()
  const seenAllowlist = new Set()
  const violations = []
  const isRepositorySource = resolve(sourceRoot) === resolve(frontendRoot, 'src')

  for (const file of await walk(sourceRoot)) {
    const path = (isRepositorySource
      ? relative(repositoryRoot, file)
      : `frontend/src/${relative(sourceRoot, file)}`
    ).replaceAll('\\', '/')
    if ((!includeTests && isTestPath(path)) || isResourcePath(path)) continue
    const source = await readFile(file, 'utf8')
    const sourceFile = ts.createSourceFile(
      file,
      source,
      ts.ScriptTarget.Latest,
      true,
      extname(file) === '.tsx' ? ts.ScriptKind.TSX : ts.ScriptKind.TS,
    )
    function visit(node) {
      if (isLiteral(node)) {
        const value = literalValue(node, sourceFile)
        if (/[А-Яа-яЁё]/u.test(value)) {
          const key = `${path}\0${fingerprint(value)}`
          const exception = acceptedExceptions.get(key)
          const allowlisted = acceptedAllowlist.get(key)
          if (exception) seenExceptions.add(key)
          else if (allowlisted) seenAllowlist.add(key)
          else if (!isMachineContract(node, value) && !isTelemetry(node)) {
            const position = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile))
            violations.push({ path, line: position.line + 1, value, fingerprint: fingerprint(value) })
          }
        }
      }
      ts.forEachChild(node, visit)
    }
    visit(sourceFile)
  }

  const staleAllowlist = [...acceptedAllowlist]
    .filter(([key]) => !seenAllowlist.has(key))
    .map(([, entry]) => entry)
  return { violations, staleAllowlist, seenExceptions: seenExceptions.size }
}

async function main() {
  const result = await scanUserFacingText()
  if (result.violations.length || result.staleAllowlist.length) {
    for (const violation of result.violations) {
      process.stderr.write(`${violation.path}:${violation.line}: user-facing Cyrillic literal: ${violation.value}\n`)
    }
    for (const entry of result.staleAllowlist) {
      process.stderr.write(`${entry.path}: stale user-facing-text allowlist entry ${entry.fingerprint}\n`)
    }
    process.exitCode = 1
  } else {
    process.stdout.write(`Frontend user-facing literal guard passed; ${result.seenExceptions} classified fixtures.\n`)
  }
}

if (process.argv[1] && resolve(process.argv[1]) === resolve(frontendRoot, 'scripts/check-user-facing-text.mjs')) {
  await main()
}

File: bot/scripts/check_user_facing_text.py
from __future__ import annotations

import ast
import hashlib
import json
import re
import sys
from dataclasses import dataclass
from pathlib import Path

CYRILLIC = re.compile(r"[А-Яа-яЁё]")


@dataclass(frozen=True)
class Finding:
    path: str
    line: int
    value: str
    fingerprint: str


@dataclass(frozen=True)
class ScanResult:
    violations: tuple[Finding, ...]
    stale_allowlist: tuple[dict[str, object], ...]
    seen_exceptions: int


def scan_user_facing_text(
    source_root: Path,
    repository_root: Path,
    exceptions_path: Path,
    allowlist_path: Path,
) -> ScanResult:
    exceptions = _entries(exceptions_path)
    allowlist = _entries(allowlist_path)
    accepted_exceptions = {
        _entry_key(entry): entry for entry in exceptions if str(entry["path"]).startswith("bot/")
    }
    accepted_allowlist = {
        _entry_key(entry): entry for entry in allowlist if str(entry["path"]).startswith("bot/")
    }
    seen_exceptions: set[str] = set()
    seen_allowlist: set[str] = set()
    violations: list[Finding] = []
    repository_source = source_root.resolve() == (repository_root / "bot/src/gym_crm_bot").resolve()

    for file in sorted(source_root.rglob("*.py")):
        relative_source = file.relative_to(source_root).as_posix()
        path = (
            file.relative_to(repository_root).as_posix()
            if repository_source
            else f"bot/src/gym_crm_bot/{relative_source}"
        )
        if "/resources/" in f"/{path}" or "/tests/" in f"/{path}":
            continue
        source = file.read_text()
        tree = ast.parse(source, filename=str(file))
        parents = {
            child: parent for parent in ast.walk(tree) for child in ast.iter_child_nodes(parent)
        }
        for node in ast.walk(tree):
            if isinstance(node, ast.Constant) and isinstance(node.value, str):
                if isinstance(parents.get(node), ast.JoinedStr):
                    continue
                value = node.value
            elif isinstance(node, ast.JoinedStr):
                value = _joined_value(node)
            else:
                continue
            if not CYRILLIC.search(value):
                continue
            fingerprint = _fingerprint(value)
            key = f"{path}\0{fingerprint}"
            if key in accepted_exceptions:
                seen_exceptions.add(key)
            elif key in accepted_allowlist:
                seen_allowlist.add(key)
            elif not _is_machine_contract(node, parents, value) and not _is_telemetry(
                node, parents
            ):
                violations.append(Finding(path, node.lineno, value, fingerprint))

    stale = tuple(entry for key, entry in accepted_allowlist.items() if key not in seen_allowlist)
    return ScanResult(tuple(violations), stale, len(seen_exceptions))


def _joined_value(node: ast.JoinedStr) -> str:
    parts: list[str] = []
    for part in node.values:
        if isinstance(part, ast.Constant):
            parts.append(str(part.value))
        elif isinstance(part, ast.FormattedValue):
            placeholder = "{" + ast.unparse(part.value)
            if part.conversion != -1:
                placeholder += "!" + chr(part.conversion)
            if part.format_spec is not None:
                placeholder += ":" + ast.unparse(part.format_spec).removeprefix("f").strip("'\"")
            parts.append(placeholder + "}")
    return "".join(parts)


def _is_machine_contract(node: ast.AST, parents: dict[ast.AST, ast.AST], value: str) -> bool:
    if value.startswith("/"):
        return True
    current: ast.AST | None = node
    while current is not None:
        if isinstance(current, ast.Assign | ast.AnnAssign):
            targets = current.targets if isinstance(current, ast.Assign) else [current.target]
            names = " ".join(ast.unparse(target).lower() for target in targets)
            return any(token in names for token in ("code", "route", "path", "callback", "command"))
        if isinstance(current, ast.keyword) and current.arg:
            return any(
                token in current.arg.lower()
                for token in ("code", "route", "path", "callback", "command")
            )
        current = parents.get(current)
    return False


def _is_telemetry(node: ast.AST, parents: dict[ast.AST, ast.AST]) -> bool:
    current: ast.AST | None = node
    while current is not None:
        if isinstance(current, ast.Call):
            target = ast.unparse(current.func).lower()
            return target == "print" or "logger." in target or "logging." in target
        current = parents.get(current)
    return False


def _entries(path: Path) -> list[dict[str, object]]:
    return list(json.loads(path.read_text())["entries"])


def _entry_key(entry: dict[str, object]) -> str:
    return f"{entry['path']}\0{entry['fingerprint']}"


def _fingerprint(value: str) -> str:
    return "sha256:" + hashlib.sha256(value.encode()).hexdigest()


def main() -> int:
    repository_root = Path(sys.argv[1]).resolve() if len(sys.argv) == 2 else Path.cwd().parent
    result = scan_user_facing_text(
        repository_root / "bot/src/gym_crm_bot",
        repository_root,
        repository_root
        / "scripts/harness/config/user-facing-text-inventory-index/scanner-exceptions.json",
        repository_root / "scripts/harness/config/user-facing-text-allowlist.json",
    )
    for finding in result.violations:
        print(
            f"{finding.path}:{finding.line}: user-facing Cyrillic literal: {finding.value}",
            file=sys.stderr,
        )
    for entry in result.stale_allowlist:
        print(
            f"{entry['path']}: stale user-facing-text allowlist entry {entry['fingerprint']}",
            file=sys.stderr,
        )
    if result.violations or result.stale_allowlist:
        return 1
    print(f"Bot user-facing literal guard passed; {result.seen_exceptions} classified fixtures.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
